crates/ruff_python_formatter/src/string/mod.rs

```diff
@@ -101,15 +102,53 @@ impl StringLikeExtensions for ast::StringLike<'_> {
         }
     }
 
-    fn is_multiline(&self, source: &str) -> bool {
-        match self {
-            Self::String(_) | Self::Bytes(_) => self.parts().any(|part| {
+    fn is_multiline(&self, context: &PyFormatContext) -> bool {
+        self.parts().any(|part| match part {
+            StringLikePart::String(_) | StringLikePart::Bytes(_) => {
                 part.flags().is_triple_quoted()
-                    && memchr2(b'\n', b'\r', source[self.range()].as_bytes()).is_some()
-            }),
-            Self::FString(fstring) => {
-                memchr2(b'\n', b'\r', source[fstring.range].as_bytes()).is_some()
+                    && context.source().contains_line_break(part.range())
             }
-        }
+            StringLikePart::FString(f_string) => {
+                fn contains_line_break_or_comments(
+                    elements: &ast::FStringElements,
+                    context: &PyFormatContext,
+                ) -> bool {
+                    elements.iter().any(|element| match element {
+                        ast::FStringElement::Literal(literal) => {
+                            context.source().contains_line_break(literal.range())
+                        }
+                        ast::FStringElement::Expression(expression) => {
+                            if is_f_string_formatting_enabled(context) {
+                                // Expressions containing comments can't be joined.
+                                //
+                                // Format specifiers needs to be checked as well. For example, the
+                                // following should be considered multiline because the literal
+                                // part of the format specifier contains a newline at the end
+                                // (`.3f\n`):
+                                //
+                                // ```py
+                                // x = f"hello {a + b + c + d:.3f
+                                // } world"
+                                // ```
+                                context.comments().contains_comments(expression.into())
+                                    || expression.format_spec.as_deref().is_some_and(|spec| {
+                                        contains_line_break_or_comments(&spec.elements, context)
```

Comment:
Do we need to consider debug expressions too?

Reply:
But they could be multiline, no?

Reply:
Hmm, so something like the following which cannot be flattened:
```py
aaaaaaaa = f"aaaaaaaaaa {
        aaaaaaa + bbbbbbb + cccccccc = } dddddddddd"
```
So, I think we should check if there's a debug expression and if so then check for line breaks in the expression itself.

Reply:
Yes, but I think we also have to test in the debug expression because the above has no line break in the expression, only in the debug expression but the whole f-string should be considered multiline (whether it can be flattened is irrelevant here because this is a general purpose method to determine if a string contains a hard line break)